RFC_STORE_MODIFIER.md

```diff
@@ -0,0 +1,185 @@
+
+The only guarantee React-Redux makes, is that if you use `replaceReducer` as a modification, the `storeState` on context will be patched with the new state for all children, even for the current render.
+
+The secondary goal of this RFC is to make it easier out of the box to code split reducers.
+
+(This file itself is never meant to be merged)
+
+## Summary of problem (recap of #1126)
+
+Omitted for brevity, see summary of the problem space in [#1150](https://github.com/reduxjs/react-redux/issues/1150), read early discussions in [#1126](https://github.com/reduxjs/react-redux/issues/1126)
+
+## Proposed solution
+
+Please note that this is just a rough sketch of what a solution could look like. I’d love feedback and to get a discussion going around both large issues and small. I've listed alternative solutions towards the end.
+
+My hope is that this solution strikes a balance between being immediately useful and preserving the flexibility for users/libraries to construct a root-reducer and injection techniques however they feel like.
+
+I am NOT proposing to hide the `ReactReduxContext` or disallow getting the store directly from it, merely to keep it as a non-public API for usecases not covered by this RFC.
+
+**Goals:**
+
+- Add API for adding reducers from the component tree
+- Optionally: Make it easier/safer to do other dynamic store modifications (register sagas, epics etc)
+- Do not require direct use of `store` or `ReactReduxContext` in userland
+- Flexible API that supports both simple and complex usecases well
+- Both existing and new libraries should be able to build more complex solutions on top of this API, but not have to worry about the intricacies of patching `storeState` on context when replacing reducers
+- Small surface area
+- Aligned with `react-redux` v6 approach for preparing for concurrent-safeness
+- SSR-support
+
+### Example usage
+
+(Some imports etc omitted for brevity)
+
+```jsx
+import { withModifiedStore } from 'react-redux'
+import reducer from './namesReducer'
+
+function Greeter({ currentName }) {
+  return `Hi ${currentName}!`
+}
+
+const mapStateToProps = state => ({
+  currentName: state.names.currentName
+})
+
+// Structure of this object is not prescriptive,
+// it depends entirely on the implementation of
+// modifyStore
+// In a sense this is quite like an action, in that it
+// is an object that should describe the store modification
+const modification = {
+  newReducers: {
+    names: reducer
+  }
+}
+
+export default withModifiedStore(modification)(
+  connect(mapStateToProps)(Greeter)
+)
+```
+
+```jsx
+import staticReducers from './staticReducers'
+
+// Example simple userland implementation
+let reducers = { ...staticReducers }
+function modifyStore({ newReducers }, storeProxy, state) {
```

Comment:
I am not a huge fan of having users write this function like this, perhaps we can go with something like the chaining method? 

```jsx
function storeModification1(modification, storeProxy, state) {
    ///whatever
}

function storeModification2(modification, storeProxy, state) {
    ///whatever 2
}

return (
    <Provider store={store} storeModifier={composeModifiers(storeModifier1, storeModifier2)}>...</Provider>
);
```

What do you think?

Reply:
I think I agree that chaining should be supported out of the box, though I'm not yet sure about the best API for that (a compose-function or supporting passing an array in)? Might be nice to support single functions as well for simple usecases.

The only unsolved issue I see with chaining is how to avoid clashes and collisions in the `modification`-object when multiple libs use it, do you have any ideas around that? :)

Reply:
My feeling is that `redux` already has a fairly solid concept for these types of things in the enhancer/middleware functionality.  I'm struggling to think of how modifiers would be different enough, or offer anything that the others don't already provide.

The only thing that is potentially missing is perhaps a more formal way of injecting values into middleware (i.e. running a new saga/observable) as each middleware currently has to implement this themselves and often they have their own, unique way of doing it.  If there was a standard API for pushing values into middleware than I think the `storeModifiers` concept goes away and then there is just a component required to call that API during rendering.

Just spitballing, but the middleware could _subscribe to modifications_ as part of their setup step:

```js

function createSagaMiddleware (options = {}) {
  function runSaga (saga) {
    // ...
  }

  return (middlewareAPI) => { // usually just called `store

    middlewareAPI.acceptModification('@@redux-saga/RUN_SAGA', (modifier) => runSaga(modifier.saga);

    return (next) => (action) => {
      // ...
    }
  }
}

const store = createStore(reducer, applyMiddleware(createSagaMiddleware()))

store.modify({ type: '@@redux-saga/RUN_SAGA', saga }) 
```

There a very strong similarity here between `actions`/`dispatch`/`subscribe` and `modifier`/`modify`/`acceptModification`, but it's important to note that they are different, both semantically and in the restrictions applied to them (e.g. actions should be serialisable, modifications would not need to be).

Reply:
This feels like a really interesting idea! 💡 I haven't thought about it deeply, but intuitively it makes a lot of sense.

I'm starting to think that treating "modifications to middleware" and "replacing the root reducer" as the same problem and cramming them into a single API might be a mistake after all.. There is definitely worth in doing them both from the same utility-component/HOC (`<StoreModifier>` for example), but should we really think about them as the same kind of problem implementation-wise? Also see [this comment](https://github.com/reduxjs/react-redux/pull/1154#issuecomment-451144369).